src/transformStateWithClones.js

```diff
@@ -7,7 +7,49 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const arrOfStates = [];
+
+  arrOfStates.push(state);
+
+  for (let i = 0; i < actions.length; i++) {
+    if (actions[i].type === 'addProperties') {
+      const stateCopy = { ...arrOfStates[arrOfStates.length - 1] };
+
+      for (const action in actions[i].extraData) {
+        stateCopy[action] = actions[i].extraData[action];
+      }
+
+      arrOfStates.push(stateCopy);
+    }
+
+    if (actions[i].type === 'clear') {
+      const stateCopy = { ...arrOfStates[arrOfStates.length - 1] };
+
+      for (const key in stateCopy) {
+        delete stateCopy[key];
+      }
+
+      arrOfStates.push(stateCopy);
+    }
+
+    if (actions[i].type === 'removeProperties') {
+      const stateCopy = { ...arrOfStates[arrOfStates.length - 1] };
+
+      for (const key in stateCopy) {
+        for (const elem of actions[i].keysToRemove) {
+          if (elem === key) {
+            delete stateCopy[key];
```

Comment:
you can use elem instead of key. If elem exists in the stateCopy, it will remove. If not - not